CI: Strongly flaky test on Windows

## Describe the bug

https://github.com/strictdoc-project/strictdoc/actions/runs/26779888295/job/78951092186

## To Reproduce

<!--

Steps to reproduce the behavior:

1. Go to '...'
2. Click on '....'
3. Scroll down to '....'
4. See error

-->

## Log

```
tests\end2end\static_html_search\02_regex_special_character_search_on_both_server_and_in_static_html\test_case.py:36: in _test_common_sequence

    self._test_regex_special_character_query(screen_project_index)

tests\end2end\static_html_search\02_regex_special_character_search_on_both_server_and_in_static_html\test_case.py:45: in _test_regex_special_character_query

    screen_project_index.assert_search_results(1, 5, 15)

tests\end2end\helpers\screens\project_index\screen_project_index.py:207: in assert_search_results

    self.test_case.assert_text("Results:")

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\seleniumbase\fixtures\base_case.py:10619: in assert_text

    self.wait_for_text_visible(text, selector, by=by, timeout=timeout)

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\seleniumbase\fixtures\base_case.py:10428: in wait_for_text_visible

    return page_actions.wait_for_text_visible(

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\seleniumbase\fixtures\page_actions.py:642: in wait_for_text_visible

    timeout_exception(TextNotVisibleException, message)

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _



exception = <class 'seleniumbase.common.exceptions.TextNotVisibleException'>

message = 'Expected text substring {Results:} for {body} was not visible after 20 seconds!'



    def timeout_exception(exception, message):

        exc, msg = shared_utils.format_exc(exception, message)

>       raise exc(msg)

E       seleniumbase.common.exceptions.TextNotVisibleException: 

E        Expected text substring {Results:} for {body} was not visible after 20 seconds!



C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\seleniumbase\fixtures\page_actions.py:270: TextNotVisibleException

- generated xml file: D:\a\strictdoc\strictdoc\build\test_reports\tests_end2end.pytest.junit.xml -

----------- Latest Logs dir: D:\a\strictdoc\strictdoc\latest_logs\ ------------

=========================== short test summary info ===========================

FAILED tests/end2end/static_html_search/02_regex_special_character_search_on_both_server_and_in_static_html/test_case.py::Test::test_search_on_server - seleniumbase.common.exceptions.TextNotVisibleException: 

 Expected text substring {Results:} for {body} was not visible after 20 seconds!

!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!

================== 1 failed, 79 passed in 800.03s (0:13:20) ===================

Error: Process completed with exit code 1.
```

## Expected behavior

<!-- A clear and concise description of what you expected to happen. -->

## Screenshots

<!-- If applicable, add screenshots to help explain your problem. -->

## Environment

<!--

 - OS: (e.g. Ubuntu Linux 22.04, macOS 12.6.3 (21G419), Windows 11)
 - Browser: (e.g. Chrome, Safari)
 - Version: (e.g. 22)

-->

## Additional context

<!-- Add any other context about the problem here. -->

--- a/developer/tasks/20260607_flaky_ci_test_in_static_html_search/task.md
+++ b/developer/tasks/20260607_flaky_ci_test_in_static_html_search/task.md
@@ -1,0 +1,43 @@
+# 2026-06-07 Fix the flaky test in Static HTML Search (Windows)
+
+## WHAT
+
+Fix the following flaky tests:
+
+- `01_search_on_both_server_and_in_static_html`
+- `02_regex_special_character_search_on_both_server_and_in_static_html`
+
+We have been observing these two tests to fail randomly, and we opened a bug report to track this:
+
+https://github.com/strictdoc-project/strictdoc/issues/2872
+
+## WHY
+
+We should have no flaky tests as they slow down the development and compromise the test suite integrity.
+
+## HOW
+
+The common flaky point is the Static HTML Search startup sequence.
+
+Both tests type into the search input immediately after the Project Index page
+is loaded. The page's `document.readyState` can already be complete while the
+generated static search index is still being loaded asynchronously. If the first
+input event is handled before the index is ready, the search UI returns without
+showing results and does not replay the input automatically.
+
+The fix is to publish an explicit startup signal from
+`static_html_search.js`:
+
+- `StrictDoc.search.isReady` stores the durable readiness state.
+- `StrictDoc.bus` publishes the `static-html-search:ready` event when the index
+  and node lookup are ready.
+
+The End-to-End Project Index screen helper waits for the usable search payload
+before it types into the Static HTML Search input. It checks the current state,
+listens for the ready event, and polls for the `index` and `nodesByMid` data so
+that a missed event or a temporary refresh state does not make the test flaky.
+This creates an intentional link between the flaky tests and the startup
+behavior they exercise.
+
+Follow the Developer Guide's newspaper style when placing helper methods:
+public screen actions/assertions stay above private `_` helpers.

--- a/strictdoc/export/html/_static/static_html_search.js
+++ b/strictdoc/export/html/_static/static_html_search.js
@@ -577,6 +577,15 @@
   // Search index initialization
   // =========================================================================
 
+  function setSearchIndexReady(isReady) {
+    strictDocSearch.isReady = isReady;
+    if (isReady) {
+      window.StrictDoc.bus.emit(
+        window.StrictDoc.events.STATIC_HTML_SEARCH_READY
+      );
+    }
+  }
+
   // Open the IndexedDB database that caches the generated search index.
   function openSearchIndexDB(name, version = 1) {
     return new Promise((resolve, reject) => {
@@ -693,6 +702,7 @@
 
       refreshInProgress = true;
       try {
+        setSearchIndexReady(false);
         await loadSearchIndexFromScript(pathToSearchIndex, true);
         await saveCurrentSearchIndexToDB({
           dbName,
@@ -700,6 +710,7 @@
           timestampMeta,
           searchData,
         });
+        setSearchIndexReady(true);
       } catch (refreshError) {
         console.error(
           "Search: Failed to refresh search index after Turbo stream update:",
@@ -845,6 +856,11 @@
   const { userinput } = dom;
   const documentLevel = parseInt(meta.documentLevel, 10);
 
+  // E2E tests rely on this startup event to type only after the async
+  // generated search index has been loaded.
+  window.StrictDoc.events.STATIC_HTML_SEARCH_READY =
+    "static-html-search:ready";
+
   const searchResultsView = new SearchResultsView(dom, {
     userinput,
     searchData: strictDocSearch,
@@ -856,6 +872,7 @@
     searchResultsView,
   });
   searchInputController.attachEventListeners();
+  setSearchIndexReady(false);
 
   // Defer search index initialization until the page and generated assets are ready.
   window.addEventListener("load", async () => {
@@ -873,5 +890,8 @@
       timestampMeta,
       searchData: strictDocSearch,
     });
+    if (strictDocSearch.index && strictDocSearch.nodesByMid) {
+      setSearchIndexReady(true);
+    }
   });
 })();

--- a/tests/end2end/helpers/screens/project_index/screen_project_index.py
+++ b/tests/end2end/helpers/screens/project_index/screen_project_index.py
@@ -27,6 +27,14 @@ from tests.end2end.helpers.screens.traceability_matrix.screen_requirements_cover
 )
 from tests.end2end.helpers.screens.tree_map.tree_map import Screen_TreeMap
 
+# Windows CI can be very slow when initializing the browser and loading
+# generated static assets, so this must be longer than the usual 10s.
+STATIC_HTML_SEARCH_READY_TIMEOUT_SECONDS = 30
+STATIC_HTML_SEARCH_READY_TIMEOUT_MS = (
+    STATIC_HTML_SEARCH_READY_TIMEOUT_SECONDS * 1000
+)
+STATIC_HTML_SEARCH_READY_POLL_INTERVAL_MS = 100
+
 
 class Screen_ProjectIndex(Screen):  # pylint: disable=invalid-name
     def __init__(self, test_case: BaseCase) -> None:
@@ -175,6 +183,8 @@ class Screen_ProjectIndex(Screen):  # pylint: disable=invalid-name
     def do_enter_search_query(self, search_query: str) -> None:
         input_xpath = "//input[@data-testid='static-html-search-input']"
 
+        self._wait_until_static_html_search_ready()
+
         # We simulate a user typing the supplied field_value.
         self.test_case.type(input_xpath, search_query, by=By.XPATH)
 
@@ -207,6 +217,90 @@ class Screen_ProjectIndex(Screen):  # pylint: disable=invalid-name
         self.test_case.assert_text("Results:")
         self.test_case.assert_text(f"{range_start}–{range_end}")
         self.test_case.assert_text(f"from {total}")
+
+    def _wait_until_static_html_search_ready(self) -> None:
+        self.test_case.assert_element(
+            "//input[@data-testid='static-html-search-input']",
+            by=By.XPATH,
+        )
+        # These tests cover startup behavior: typing before the async static
+        # search index is ready can drop the first input event.
+        self.test_case.driver.set_script_timeout(
+            STATIC_HTML_SEARCH_READY_TIMEOUT_SECONDS
+        )
+        wait_result = self.test_case.driver.execute_async_script(
+            """
+            const searchReadyTimeoutMs = arguments[0];
+            const pollIntervalMs = arguments[1];
+            const done = arguments[arguments.length - 1];
+            const eventName =
+              window.StrictDoc?.events?.STATIC_HTML_SEARCH_READY ||
+              "static-html-search:ready";
+            let timeout = null;
+            let polling = null;
+
+            function collectSearchStatus() {
+              const searchData = window.StrictDoc && window.StrictDoc.search;
+              const hasIndex = !!(searchData && searchData.index);
+              const hasNodesByMid = !!(searchData && searchData.nodesByMid);
+              return {
+                isUsable: hasIndex && hasNodesByMid,
+                hasStrictDoc: !!window.StrictDoc,
+                hasSearchData: !!searchData,
+                isReadyFlag: !!(searchData && searchData.isReady),
+                hasIndex: hasIndex,
+                hasNodesByMid: hasNodesByMid,
+                documentReadyState: document.readyState,
+                location: window.location.href,
+              };
+            }
+
+            function resolveIfUsable() {
+              const status = collectSearchStatus();
+              if (status.isUsable) {
+                cleanup();
+                done(status);
+                return true;
+              }
+              return false;
+            }
+
+            function cleanup() {
+              if (timeout !== null) {
+                window.clearTimeout(timeout);
+              }
+              if (polling !== null) {
+                window.clearInterval(polling);
+              }
+              document.removeEventListener(eventName, onReady);
+            }
+
+            function onReady() {
+              resolveIfUsable();
+            }
+
+            if (resolveIfUsable()) {
+              return;
+            }
+
+            timeout = window.setTimeout(() => {
+              cleanup();
+              done(collectSearchStatus());
+            }, searchReadyTimeoutMs);
+
+            polling = window.setInterval(
+              resolveIfUsable,
+              pollIntervalMs
+            );
+            document.addEventListener(eventName, onReady, { once: true });
+            """,
+            STATIC_HTML_SEARCH_READY_TIMEOUT_MS,
+            STATIC_HTML_SEARCH_READY_POLL_INTERVAL_MS,
+        )
+        assert wait_result["isUsable"], (
+            "Static HTML search index was not ready before entering a query. "
+            f"Search status: {wait_result}."
+        )
 
     #
     # Add new document

